### PR TITLE
feat: add code mods for parse5 parsing errors

### PR DESCRIPTION
**Pull request by jmsjtu**
This PR introduces code mods to fix parse5 parsing errors. 

These are the list of errors:
- `end-tag-without-matching-open-element`
- `closing-of-element-with-open-child-elements`
- `eof-in-element-that-can-contain-only-text`

The full list of `parse5` errors can be found [here](https://github.com/salesforce/lwc/issues/3258), however, only the 3 above are addressed as a majority of the errors listed in the issue are not applicable because LWC uses the `parseFragment` function.

### CODE REVIEW

File: tests/fixtures/closing-of-element-with-open-child-elements/expected/x/component/component.html [hunk 1]
```diff
@@ -0,0 +1,5 @@
+<template>
+    <section>
+        <span>hello</span>
+        <span>mr. anderson</span>
+</section></template>
```
Comment (jmsjtu) [inline tests/fixtures/closing-of-element-with-open-child-elements/expected/x/component/component.html:5]: There isn't an easy way to reformat the tags so they line up perfectly with their starting tag.

File: tests/fixtures/end-tag-without-matching-open-element/multiple/expected/x/component/component.html [hunk 2]
```diff
@@ -0,0 +1,2 @@
+<template>
+    <span>hello </span></template>
```
Comment (jmsjtu) [inline tests/fixtures/end-tag-without-matching-open-element/multiple/expected/x/component/component.html:2]: Same thing here, there isn't a straightforward way to reformat the tags to line up correctly.

File: README.md [hunk 3]
```diff
@@ -30,6 +30,7 @@ Available transforms:
 
 - [Shadow DOM to Light DOM](#shadow-dom-to-light-dom) (`shadow-to-light`)
 - [Synthetic Shadow DOM to Native Shadow DOM](#synthetic-shadow-dom-to-native-shadow-dom) (`synthetic-to-native`)
+- [HTML parsing errors](#html-parsing-errors)
```
Comment (nolanlawson) [inline README.md:33]: I think potentially we can put all template cleanup functions into a single codemod. They all have the same goal: remove warnings/errors in `.html` files.

In the context of an IDE for instance, the scenario would probably be:

1. User is editing an `.html` file
2. The IDE suggests you run a codemod on it

Plus, if we add codemods in the future for .js and .css files, then there are only three total, which seems simple to run on a whole codebase.

File: README.md [hunk 4]
```diff
@@ -81,6 +82,23 @@ Converts components from synthetic shadow to native shadow.
...
+
+#### Usage
+
+```sh
+lwc-codemod <error-name> <path>
+```
+
+#### Summary
+
+Fixes parsing HTML parsing errors generated in LWC through `parse5`
```
Comment (nolanlawson) [inline README.md:95]:
```suggestion
Fixes errors generated during LWC HTML parsing
```

File: transforms/parse5Utils.js [hunk 5]
```diff
@@ -3,49 +3,49 @@ import * as parse5Module from 'parse5'
...
+  setAttr(node, 'class', (attr.trim() + ' ' + className).trim())
 }
 
 export const replaceNode = (node, replacement) => {
-    const parentIdx = node.parentNode.childNodes.indexOf(node)
-    node.parentNode.childNodes[parentIdx] = replacement
-    replacement.parent = node.parent
+  const parentIdx = node.parentNode.childNodes.indexOf(node)
+  node.parentNode.childNodes[parentIdx] = replacement
+  replacement.parent = node.parent
```
Comment (nolanlawson) [inline transforms/parse5Utils.js:50]: Same comment in the other PR about whitespace/`standard`

**Comment by nolanlawson** [comment]
@jmsjtu Looks like you have a merge conflict? ^

File: README.md [hunk 6]
```diff
@@ -142,6 +142,15 @@ Note that for non-root `template` elements, if there aren't any valid directives
...
 
+##### HTML parsing errors
+
+Fixes errors generated during LWC HTML parsing.
+
+The following transforms are available for each errors:
+- `eof-in-element-that-can-contain-only-text`
+- `end-tag-without-matching-open-element`
+- `closing-of-element-with-open-child-elements`
+
```
Comment (jmsjtu) [inline README.md:153]: This section is nested under the `html-template-cleanup` code mod and just explains the different fixes. 

I wanted to put examples here but these aren't official HTML errors.

I think they're specific to `parse5` and I couldn't find great examples for them other than in this [repo](https://github.com/rehypejs/rehype/blob/main/packages/rehype-parse/readme.md).

**Comment by jmsjtu** [comment]
@nolanlawson Updated this PR to point to #15 as the base and it now combines the `parse5` fixes into a single codemod defined there.

File: README.md (hunk 6 above)
Comment (nolanlawson) [reply README.md:153]: Yeah parse5 gets some of these names from the spec, but other ones are just invented.

**Review by nolanlawson — APPROVED** [review]
Awesome!

File: transforms/htmlTemplateCleanup/endTagWithoutMatchingOpenElement.js [hunk 7]
```diff
@@ -0,0 +1,19 @@
+/*
+ * Copyright (c) 2021, salesforce.com, inc.
+ * All rights reserved.
+ * SPDX-License-Identifier: BSD-3-Clause
+ * For full license text, see the LICENSE file in the repo root or https://opensource.org/licenses/BSD-3-Clause
+ */
+import { serializeParse5Ast } from '../parse5Utils'
```
Comment (nolanlawson) [inline transforms/htmlTemplateCleanup/endTagWithoutMatchingOpenElement.js:7]:
```suggestion
import { serializeParse5Ast } from '../parse5Utils.js'
```

File: transforms/htmlTemplateCleanup/closingOfElementWithOpenChildElements.js [hunk 8]
```diff
@@ -0,0 +1,29 @@
+/*
+ * Copyright (c) 2021, salesforce.com, inc.
+ * All rights reserved.
+ * SPDX-License-Identifier: BSD-3-Clause
+ * For full license text, see the LICENSE file in the repo root or https://opensource.org/licenses/BSD-3-Clause
+ */
+import { walkParse5Ast, serializeParse5Ast } from '../parse5Utils'
```
Comment (nolanlawson) [inline transforms/htmlTemplateCleanup/closingOfElementWithOpenChildElements.js:7]:
```suggestion
import { walkParse5Ast, serializeParse5Ast } from '../parse5Utils.js'
```

File: transforms/htmlTemplateCleanup/eofInElementThatCanContainOnlyText.js [hunk 9]
```diff
@@ -0,0 +1,35 @@
+/*
+ * Copyright (c) 2021, salesforce.com, inc.
+ * All rights reserved.
+ * SPDX-License-Identifier: BSD-3-Clause
+ * For full license text, see the LICENSE file in the repo root or https://opensource.org/licenses/BSD-3-Clause
+ */
+import { walkParse5Ast, serializeParse5Ast } from '../parse5Utils'
```
Comment (nolanlawson) [inline transforms/htmlTemplateCleanup/eofInElementThatCanContainOnlyText.js:7]:
```suggestion
import { walkParse5Ast, serializeParse5Ast } from '../parse5Utils.js'
```

**Comment by nolanlawson** [comment]
All the missing `.js`es in the imports make me nervous. I think we should really add a sanity test to GH Actions to run the CLI directly, rather than through Jest. Otherwise we could get ✅ on our tests, publish to npm, and then it's broken.

I can do that, no worries.

**Pull request merged**